In [1]:
import numpy as np

import tensorflow as tf
from tensorflow.keras import layers, models, losses, optimizers, datasets, preprocessing, utils

In [2]:
# 구글 드라이브 마운트
from google.colab import drive

drive.mount('/content/drive') 

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [3]:
import os

drive_dir = "/content/drive/MyDrive/"
data_dir = os.path.join(drive_dir, "colab", "CNU_AI_TEAMWORK", "Term_Dataset")
model_dir = os.path.join(data_dir, "apple.h5")
data_dir

'/content/drive/MyDrive/colab/CNU_AI_TEAMWORK/Term_Dataset'

In [4]:
from keras.models import load_model
from tensorflow.keras.utils import plot_model

In [5]:
model = load_model(model_dir)

In [ ]:
plot_model(model, to_file='model_shapes.png', show_shapes=True)

In [6]:
batch_size = 32
img_h = 256
img_w = 256
num_classes = 11

In [7]:
val_test_datagen = preprocessing.image.ImageDataGenerator(
    rescale = 1.0/255,
)

In [8]:
val_test_dir = os.path.join(drive_dir, "colab", "CNU_AI_TEAMWORK", "Term_Dataset", "val_test") # 이경로를 사용할예정 여기 shoe에 데이터를 전부 넣으면됌

# val_test 안에 임의의 폴더가 있어야함 categorical로 분류할 ex) "shoe", 경로에는 안씀
# 임의의 폴더안에 이미지 133개를 모두 넣으면 됌

val_test_ds = val_test_datagen.flow_from_directory(
    directory = val_test_dir,
    target_size = (img_h, img_w),
    class_mode = 'categorical',
    # subset = 'testing',
    shuffle = False,
    batch_size = 32,
    seed=42
)

Found 288 images belonging to 1 classes.


In [12]:
predictions = model.predict(val_test_ds)
pred_labels = np.argmax(predictions, axis = -1) # We take the highest probability
pred_labels

array([ 0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,
        7,  7,  0,  7,  0,  7,  0,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,
        1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  2,  2,  2,
        2,  2,  2,  2,  2,  2,  2,  1,  2,  1,  2,  2,  2,  2,  2,  2,  2,
        2,  2,  2,  2,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3,
        3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  3,  4,  4,  4,  4,  4,  4,
        4,  4,  4,  4,  4,  4,  4,  4,  4,  4,  4,  4,  4,  4,  4,  4,  4,
        4,  5,  5,  5,  5,  5,  5,  5,  5,  5,  5,  5,  5,  5,  5,  5,  5,
        5,  5,  5,  5,  5,  5,  5,  5,  6,  6,  6,  6,  6,  6,  6,  6,  6,
        6,  6,  6,  6,  6,  6,  6,  6,  6,  6,  6,  6,  6,  6,  6,  6,  6,
        6,  6,  6,  6,  6,  6,  6,  6,  6,  6,  6,  6,  6,  6,  6,  6,  7,
        7,  7,  7,  7,  7,  7,  7,  7,  7,  7,  7,  7,  7,  7,  7,  7,  7,
        8,  8,  8,  8,  8,  8,  8,  8,  8,  8,  8,  8,  8,  8,  8,  8,  8,
        8,  8,  8,  8,  8

In [13]:
def num_to_class(file_dir): #/content/drive/MyDrive/colab/CNU_AI_TEAMWORK/Term_Dataset/val_data_emp.csv
  z = pd.read_csv(file_dir)

  for i in range(len(z)):
    if z.loc[i:i,['class']].values.astype(str)[0][0] == '0':
      z['class'][i] = 'bag'
    if z.loc[i:i,['class']].values.astype(str)[0][0] == '1':
      z['class'][i] = 'bed'
    if z.loc[i:i,['class']].values.astype(str)[0][0] == '2':
      z['class'][i] = 'chair'
    if z.loc[i:i,['class']].values.astype(str)[0][0] == '3':
      z['class'][i] = 'coffeetable'
    if z.loc[i:i,['class']].values.astype(str)[0][0] == '4':
      z['class'][i] = 'cup'
    if z.loc[i:i,['class']].values.astype(str)[0][0] == '5':
      z['class'][i] = 'kitchentools'
    if z.loc[i:i,['class']].values.astype(str)[0][0] == '6':
      z['class'][i] = 'lamp'
    if z.loc[i:i,['class']].values.astype(str)[0][0] == '7':
      z['class'][i] = 'laptop'
    if z.loc[i:i,['class']].values.astype(str)[0][0] == '8':
      z['class'][i] = 'LivingSofa'
    if z.loc[i:i,['class']].values.astype(str)[0][0] == '9':
      z['class'][i] = 'pot'
    if z.loc[i:i,['class']].values.astype(str)[0][0] == '10':
      z['class'][i] = 'shoe'

  z.to_csv(file_dir, index=False)

In [14]:
import pandas as pd
predictions = model.predict(val_test_ds)
pred_labels = np.argmax(predictions, axis = -1) # We take the highest probability
pred_labels

submission = pd.read_csv('/content/drive/MyDrive/colab/CNU_AI_TEAMWORK/Term_Dataset/test_data_emp.csv') # csv 파일 위치
submission['class'] = np.argmax(predictions, axis=-1) # 이게 정상적으로 안된다면? 안될 확률이 매우높음
submission.to_csv('/content/drive/MyDrive/colab/CNU_AI_TEAMWORK/Term_Dataset/apple.csv', index=False)
num_to_class('/content/drive/MyDrive/colab/CNU_AI_TEAMWORK/Term_Dataset/apple.csv')

KeyboardInterrupt: ignored

288개의 데이터가 있다고 가정

csv파일은 정렬이 안되있음 

하지만 구글 drive는 정렬이 되어있음 

csv파일을 정렬한 상태로 읽어
삽입을 한 임의의 csv파일을 생성하고

정렬된 상태

2개의 csv파일에서 원래의 csv파일의 순서에 맞게 재조정


In [15]:
import pandas as pd
submission = pd.read_csv('/content/drive/MyDrive/colab/CNU_AI_TEAMWORK/Term_Dataset/test_data_emp.csv') # test csv 파일 위치 나중에 test.emp.csv? 이게 올 예정 
val_list = submission.file_name.tolist()
val_list_sorted = sorted(val_list)
val_list_sorted

submission['file_name'] = val_list_sorted
submission['class'] = np.argmax(predictions, axis=-1)
submission.to_csv('/content/drive/MyDrive/colab/CNU_AI_TEAMWORK/Term_Dataset/apple_temp.csv', index=False) # 정렬이 되어있는 csv 파일
num_to_class('/content/drive/MyDrive/colab/CNU_AI_TEAMWORK/Term_Dataset/apple_temp.csv')
submission['file_name'] = val_list
submission.to_csv('/content/drive/MyDrive/colab/CNU_AI_TEAMWORK/Term_Dataset/apple.csv', index=False) # 제출해야할 정렬이 되어있지 않은 csv 파일

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:6: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  
/usr/local/lib/python3.7/dist-packages/pandas/core/indexing.py:1732: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self._setitem_single_block(indexer, value, name)


정렬되있는거와 정렬안되있는거를 맞춰줌

In [16]:
real_submission = pd.read_csv('/content/drive/MyDrive/colab/CNU_AI_TEAMWORK/Term_Dataset/apple.csv') # 진짜 제출할것 정렬이 안되있음
temp_submission = pd.read_csv('/content/drive/MyDrive/colab/CNU_AI_TEAMWORK/Term_Dataset/apple_temp.csv') # 진짜 제출할것 정렬이 되있음

file_name_list = real_submission.file_name.tolist()
temp_file_name_list = temp_submission.file_name.tolist()
file_name_list

real_class_list =[]

for i in range(len(real_submission)):
  find_image = file_name_list[i]
  found = temp_file_name_list.index(find_image)
  category_class = temp_submission['class'][found]
  real_class_list.append(category_class)  

real_submission['class'] = real_class_list
real_submission.to_csv('/content/drive/MyDrive/colab/CNU_AI_TEAMWORK/Term_Dataset/apple.csv', index=False)
num_to_class('/content/drive/MyDrive/colab/CNU_AI_TEAMWORK/Term_Dataset/apple.csv')

real_submission

Unnamed: 0     file_name   class
0             0  img00018.png  laptop
1             1  img00024.png     bag
2             2  img00019.png  laptop
3             3  img00022.png     bag
4             4  img00023.png  laptop
..          ...           ...     ...
283         283  img00288.png    shoe
284         284  img00261.png    shoe
285         285  img00275.png    shoe
286         286  img00274.png    shoe
287         287  img00260.png    shoe

[288 rows x 3 columns]

In [ ]:
# import pandas as pd
# import random

# submission = pd.read_csv('/content/drive/MyDrive/colab/CNU_AI_TEAMWORK/Term_Dataset/val_data_emp.csv') # test csv 파일 위치
# val_list = submission.file_name.tolist()
# random.shuffle(val_list)
# submission['file_name'] = val_list
# submission.to_csv('/content/drive/MyDrive/colab/CNU_AI_TEAMWORK/Term_Dataset/val_data_emp_test.csv', index=False) # shuffled data
